core/src/idx/trees/hnsw/mod.rs

```diff
@@ -35,97 +41,170 @@ impl HnswSearch {
 		}
 	}
 }
+
+#[revisioned(revision = 1)]
+#[derive(Default, Serialize, Deserialize)]
+pub(super) struct HnswState {
+	enter_point: Option<ElementId>,
+	next_element_id: ElementId,
+	layer0: LayerState,
+	layers: Vec<LayerState>,
+}
+
+impl VersionedStore for HnswState {}
+
 struct Hnsw<L0, L>
 where
-	L0: DynamicSet<ElementId>,
-	L: DynamicSet<ElementId>,
+	L0: DynamicSet,
+	L: DynamicSet,
 {
+	ikb: IndexKeyBase,
+	state_key: Key,
+	state: HnswState,
 	m: usize,
 	efc: usize,
 	ml: f64,
 	layer0: HnswLayer<L0>,
 	layers: Vec<HnswLayer<L>>,
-	enter_point: Option<ElementId>,
 	elements: HnswElements,
 	rng: SmallRng,
 	heuristic: Heuristic,
 }
 
-pub(super) type ElementId = u64;
+pub(crate) type ElementId = u64;
 
 impl<L0, L> Hnsw<L0, L>
 where
-	L0: DynamicSet<ElementId>,
-	L: DynamicSet<ElementId>,
+	L0: DynamicSet,
+	L: DynamicSet,
 {
-	fn new(p: &HnswParams) -> Self {
+	fn new(ikb: IndexKeyBase, p: &HnswParams) -> Self {
 		let m0 = p.m0 as usize;
+		let state_key = ikb.new_hs_key();
 		Self {
+			state_key,
+			state: Default::default(),
 			m: p.m as usize,
 			efc: p.ef_construction as usize,
 			ml: p.ml.to_float(),
-			enter_point: None,
-			layer0: HnswLayer::new(m0),
+			layer0: HnswLayer::new(ikb.clone(), 0, m0),
 			layers: Vec::default(),
-			elements: HnswElements::new(p.distance.clone()),
+			elements: HnswElements::new(ikb.clone(), p.distance.clone()),
 			rng: SmallRng::from_entropy(),
 			heuristic: p.into(),
+			ikb,
+		}
+	}
+
+	async fn check_state(&mut self, tx: &Transaction) -> Result<(), Error> {
+		// Read the state
+		let st: HnswState = if let Some(val) = tx.get(self.state_key.clone(), None).await? {
```

Comment:
@emmanuel-keller Hey! 👋 Just trying to understand the perf implication here; Am I correct that you need to read the HNSW State from the state key for every HNSW index used in a transaction?

Also, can I safely assume you must include the write to the state key for every index being updated within a transaction, so it will be a write hotspot?

Reply:
You're right, we are updating the state (and increasing the version of the affected layers) on every write.
At the end of the day, for HNSW (and indexes in general), having multiple writers is an anti-pattern.
We have to ensure that at least on the same SurrealDB instance, concurrency is well-managed on the indexing side so that no concurrent write can easily trigger transaction conflicts.  
That's a fair point! Ideally, this PR will go into the beta so we can have feedback on the HNSW persistence.
I will specifically address this point during the beta process in an upcoming PR.